### ISSUE TITLE
hapOut sporadically gets into a fail state caused by WEBLOG thread safety

### ISSUE THREAD

**Issue by frankonski** [HomeSpan/HomeSpan#899]
While trying to figure out why hitting weblogs was not always sending the full HTML page, and why sometimes the Home app returns "No response", I ended up adding a bit of debug in `HAPClient::getStatusURL()` as such:
```
  hapOut << "</body></html>\n";
  hapOut.flush();
  // Let's see if there was an error flushing the stream
  if(!hapOut.good()) {
    LOG0("Error flushing stream. (eofbit: %d, failbit: %d, badbit: %d)\n", hapOut.eof(), hapOut.fail(), hapOut.bad());
  }
```
And the outcome I get, from time to time, is
```
Error flushing stream. (eofbit: 0, failbit: 1, badbit: 1)
```
No idea why at the moment, and no TLS code, just standard weblogs.

And there does not seem to be a way to recover from this other than reboot the device.

Hitting Refresh in Chrome gives 
```
192.168.1.106 didn’t send any data.
ERR_EMPTY_RESPONSE
```
and turning L2 logs, I get this:
```
=======================================
** Client #2 Connected (773 sec): 192.168.1.116

<<<<<<<<< 192.168.1.116 <<<<<<<<<
GET /status HTTP/1.1
Host: 192.168.1.106
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9,fr;q=0.8
------------ END BODY! ------------

>>>>>>>>>> 192.168.1.116 >>>>>>>>>>
Error flushing stream. (eofbit: 0, failbit: 1, badbit: 1)
------------ SENT! --------------
=======================================
** Client #2 Connected (773 sec): 192.168.1.116

** Client #2 DISCONNECTED (773 sec)
```

**Comment by frankonski** [comment]
`hapOut` is global right? Wouldn't it need to be protected against concurrent access, like when a controller requests an update from HomeSpan, and when a user hits the weblog?

Wouldn't you have two distinct requests flushing and/or calling `setHapClient()` ?

**Comment by HomeSpan** [comment]
Hmmm.  I've not encountered this.

Note there is no concurrent access (this is why things like the AsyncWebServer does not play well with HomeSpan).   Calls to the web log are treated the same as calls from a Controller.  Each is processed to completion before the next request is handled.

**Comment by frankonski** [comment]
I can't find a way to reproduce this.. sporadic and random. Recently, sometimes almost immediately after reflash or reboot.

**Comment by frankonski** [comment]
And not related to `flush()` either.
```
Before flushing stream. (goodbit: 0, eofbit: 0, failbit: 1, badbit: 1)
Error flushing stream. (eofbit: 0, failbit: 1, badbit: 1)
```

**Comment by frankonski** [comment]
A bit more logging, I now got to this:
```
iostate 10: (goodbit: 1, eofbit: 0, failbit: 0, badbit: 0)
Before flushing stream. (goodbit: 0, eofbit: 0, failbit: 1, badbit: 1)
Error flushing stream. (eofbit: 0, failbit: 1, badbit: 1)
```
with this code
```
  LOG0("iostate 10: (goodbit: %d, eofbit: %d, failbit: %d, badbit: %d)\n", hapOut.good(), hapOut.eof(), hapOut.fail(), hapOut.bad());
  if(homeSpan.webLog.maxEntries>0){
    // I'm not reaching 500 entries yet.
    // ...
  }
 
  hapOut << "</body></html>\n";
  
  LOG0("Before flushing stream. (goodbit: %d, eofbit: %d, failbit: %d, badbit: %d)\n", hapOut.good(), hapOut.eof(), hapOut.fail(), hapOut.bad());
```
Now I wonder something. HTTP requires the use of `\r\n`, not just `\n`.. but still, if the buffer hasn't been flushed, can't expect the socket to be closed. I will try with `\r\n` for a while.

**Comment by frankonski** [comment]
I didn't read correctly and spoke to quickly.
```
iostate 10: (goodbit: 1, eofbit: 0, failbit: 0, badbit: 0)
iostate 11: (goodbit: 0, eofbit: 0, failbit: 1, badbit: 1)
Before flushing stream. (goodbit: 0, eofbit: 0, failbit: 1, badbit: 1)
Error flushing stream. (eofbit: 0, failbit: 1, badbit: 1)
```
I'm narrowing it down to something happening inside the IF `if(homeSpan.webLog.maxEntries>0){` as I added "iostate 11" right there:
```
  LOG0("iostate 11: (goodbit: %d, eofbit: %d, failbit: %d, badbit: %d)\n", hapOut.good(), hapOut.eof(), hapOut.fail(), hapOut.bad());
  hapOut << "</body></html>\n\r\n";
```

**Comment by frankonski** [comment]
The HTML source when this happens seems to always end like this:
```
... top is ok ...
<p></p><table class=tab2><tr><th>Entry</th><th>Up Time</th><th>Log Time</th><th>Client</th><th>Message</th></tr>
<tr><td>9</td><td>0:00:00:06</td><td>Sun Aug  4 17:05:04 2024</td><td>0.0.0.0</td><td>Trouble light ON.</td></tr>
<tr><td>8</td><td>0:00:00:06</td><td>Sun Aug  4 17:05:04 2024</td><td>0.0.0.0</td><td>Bureau: Armed for Away </td></tr>
<tr><td>7</td><td>0:00:00:06</td><td>Sun Aug  4 17:05:04 2024</td><td>0.0.0.0</td><td>Keybus reconnected. Trouble light OFF.</td></tr>
<tr><td>6</td><td>0:00:00:06</td><td>Sun Aug  4 17:05:04 2024</td><td>0.0.0.0</td><td>Country code: US, country maxTxPower: 30, currentTxPower: 19.50 (max:19.50), policy: AUTO. Setting country code.</td></tr>
<tr><td>5</td><td>0:00:00:00</td><td>Unknown</td><td>
```
As seen above, the last `<td>` is where we'd normally see the IP address, like if `homeSpan.webLog.log[index].clientIP.c_str()` is somewhat corrupted...
More to come.

**Comment by frankonski** [comment]
That last comment above led to nowhere.
However, one thing for sure, I can easily constantly (but randomly) reproduce it when using the original HS code `hapOut << "</body></html>\n";`, and the issue cannot be reproduce when sending CRLF on a new line as per the RFC: `hapOut << "</body></html>\n\r\n";`

Steps to reproduce
- I have 5-6 lines that gets logged with `WEBLOG` level prior to wifi being established (if that makes any difference).
- Have Chrome open to http://<ip>/status
- Reboot the device
- Wait for HAP server to be ready by seeing this in the serial console: `Starting HAP Server on port 80...`
- Hit refresh constantly. If the issue has not reproduced itself within ~20 refresh, reboot and try again.

**Comment by frankonski** [comment]
I added this:
```
      if (!homeSpan.webLog.log[index].clientIP.c_str()) {
        LOG0("NO IP !!!\n");
      }
```
And this is what causes the failbit to be set. I got the logs. Looking at the implementation of `operator<<` of `ostream`, it does set the failbit when the string is null.

So the problematic line is this:
```
      hapOut << "<tr><td>" << i+1 << "</td><td>" << uptime << "</td><td>" << clocktime << "</td><td>" << homeSpan.webLog.log[index].clientIP.c_str() << "</td><td>" << homeSpan.webLog.log[index].message << "</td></tr>\n";
```
.. when `homeSpan.webLog.log[index].clientIP.c_str()` is null.

The questions now is _why is clientIP null on some occasions_?

**Comment by frankonski** [comment]
I don't understand. In `HomeSpan.cpp`, in `vLog`, I added serial logging like such:
```
    log[index].message=(char *)HS_REALLOC(log[index].message, strlen(buf) + 1);
    strcpy(log[index].message, buf);
    
    Serial.print("Last IP: ");
    Serial.println(homeSpan.lastClientIP);

    log[index].clientIP=homeSpan.lastClientIP;  
    nEntries++;
```
and in `HAP.cpp`, in `getStatusURL()`, I added this logging:
```
      if(homeSpan.webLog.log[index].clockTime.tm_year>0)
        strftime(clocktime,sizeof(clocktime),"%c",&homeSpan.webLog.log[index].clockTime);
      else
        sprintf(clocktime,"Unknown");        
      
      Serial.print("--- item - IP: ");
      Serial.println(homeSpan.webLog.log[index].clientIP);
      hapOut << "<tr><td>" << i+1 << "</td><td>" << uptime << "</td><td>" << clocktime << "</td><td>" << homeSpan.webLog.log[index].clientIP.c_str() << "</td><td>" << homeSpan.webLog.log[index].message << "</td></tr>\n";
```
and interestingly, this is what gets logged:
```
Last IP: 0.0.0.0
Last IP: 0.0.0.0
Last IP: 0.0.0.0
Last IP: 0.0.0.0
Last IP: 0.0.0.0
Last IP: 0.0.0.0
Last IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP:
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
```

The IP that gets clipped comes from a `WEBLOG` I do in a `setWifiCallback` callback method.

**Comment by frankonski** [comment]
Still chasing down. In `getStatusURL()` :
```
      Serial.print("--- item - IP: ");
      Serial.println(homeSpan.webLog.log[index].clientIP);
      if (!homeSpan.webLog.log[index].clientIP) {
        Serial.println("***  IP is null!");
      }
      hapOut << "<tr><td>" << i+1 << "</td><td>" << uptime << "</td><td>" << clocktime << "</td><td>" << homeSpan.webLog.log[index].clientIP.c_str() << "</td><td>" << homeSpan.webLog.log[index].message << "</td></tr>\n";
```
I get this:
```
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP:
***  IP is null!
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
--- item - IP: 0.0.0.0
```

I also added checks for all allocations / assignations of `clientIP` to check if _clientIP = something_ completes successfully, and there is no error. clientIP is never null after assignation, but it gets to null in the function.

Any advice?

**Comment by frankonski** [comment]
.. there is a threading issue going on. See where it logs **"About to copy 3"** twice.
```
Starting HAP Server on port 80...

About to copy 3
   Copy done, 3 entries

ClientIP: 0.0.0.0
About to copy 3
   Copy done, 4 entries
About to copy 5
   Copy done, 5 entries
About to copy 6
   Copy done, 6 entries
About to copy 7
   Copy done, 7 entries
About to copy 8
   Copy done, 8 entries
About to copy 9
   Copy done, 9 entries
```
This is the modified code:
```
    Serial.printf("About to copy %d\n", nEntries + 1);
    log[index].clientIP=homeSpan.lastClientIP;
    Serial.printf("   Copy done, %d entries\n", nEntries + 1);
    nEntries++;
```

That `homeSpan.setWifiCallback()` callback, does it get executed by a different thread that the main loop?

**Comment by frankonski** [comment]
I nailed it. `WEBLOG` / `vLog` is not thread safe.
Using this code as a proof:
```
    Serial.printf("About to copy %d - task name: %s\n", nEntries + 1, pcTaskGetName(NULL));
    log[index].clientIP=homeSpan.lastClientIP;
    Serial.printf("   Copy done, %d - task name: %s\n", nEntries + 1, pcTaskGetName(NULL));
```
produces the output:
```
About to copy 1 - task name: loopTask
   Copy done, 1 - task name: loopTask
WiFi Connected!  IP Address = 192.168.80.106
About to copy 2 - task name: loopTask
   Copy done, 2 - task name: loopTask

Starting MDNS...

<removed>

Starting HAP Server on port 80...

About to copy 3 - task name: timeSeverTaskHa
   Copy done, 3 - task name: timeSeverTaskHa
About to copy 3 - task name: loopTask
   Copy done, 4 - task name: loopTask
About to copy 5 - task name: loopTask
About to copy 5 - task name: timeSeverTaskHa
   Copy done, 5 - task name: timeSeverTaskHa
   Copy done, 6 - task name: loopTask
About to copy 7 - task name: loopTask
   Copy done, 7 - task name: loopTask
About to copy 8 - task name: loopTask
   Copy done, 8 - task name: loopTask
About to copy 9 - task name: loopTask
   Copy done, 9 - task name: loopTask
About to copy 10 - task name: loopTask
   Copy done, 10 - task name: loopTask
```
Again, you can see there is a copy **twice** to index 3, which mean index 4 will remain null.

Now the task name shows `timeSeverTaskHa` ... About a week-or-so ago, I setup my local NTP server. So instead of reaching out to the web for an NTP server, all this local, so NTP response happens really fast, causing HomeSpan to log its own WEBLOG in its own thread called `timeSeverTaskHandle`.

I'll see if I can submit a PR to get `WEBLOG` / `vLog` thread safe, unless @HomeSpan you prefer tackling it?

**Comment by HomeSpan** [comment]
I finally had a chance to explore this issue and you are indeed correct that there is a problem with the thread safety.  If you were running HomeSpan with autoPolling you would find that any calls to WEBLOG you made from outside of the polling thread, such as using WEBLOG in the main Arduino `loop()` method, could cause a thread issue if you were simultaneously trying to read the Web Log in a browser.

But even if did not use autoPolling, there is still a separate thread that calls WEBLOG behind the scenes - namely the background task that sets the time via an NTP server.  This background tasks terminates after the time is read, but as you've found above, it still can cause a thread collision.

Both issues can be addressed by making the Web Log code thread-safe.  After doing some research I found there is a standardized way in C++ of using a mutex to solve the problem of inter-locking separate functions that can read and write to the same shared variables.  The key is using a shared mutex and instead of manually locking and unlocking, you simply create a shared_lock or unique_lock object whenever you want to protect a piece of code.  As soon as the code goes out of scope, the object is destroyed and upon destruction it releases the lock.  The other advantage to using this method is that it allows multiple threads to read at the same time, but only one thread can write at any given time.  Since the mutex is shared, the system knows that it should block all threads from reading if anything is being written, and it should wait to start writing until any threads that are reading are done.

The changes to the code were very minimal (just a few extra lines) and I've uploaded to the dev branch.  Please give this a try and let me know if it solved the issue in this thread (no pun intended).  I tested myself using autoPolling with a log of extra weblog writes coming from the main Arduino `loop()` every 100ms and things seems to work as expected.

NOTE that the issue you have outlined in your other thread (#903) may or may not be related.  Once we tackle this we can explore the next.

Also, if this works I will add similar shared mutex locks to all other entry points, such as getVal and setVal to ensure you can read and write to Characteristics even from an external thread.  I knew there was a good reason I labeled autoPolling as experimental when initially released.  Maybe after all this is solved, I'll take off the "experimental" label.

**Comment by HomeSpan** [comment]
Feature is now available in latest release v2.1.0.
